src/lib/distances.jl

```diff
@@ -59,26 +59,25 @@ end
 @adjoint function colwise(s::Euclidean, x::AbstractMatrix, y::AbstractMatrix)
   d = colwise(s, x, y)
   return d, function (Δ::AbstractVector)
-    x̄ = (Δ ./ d)' .* (x .- y)
+    x̄ = (Δ ./ (d .+ eps(eltype(d))))' .* (x .- y)
     return nothing, x̄, -x̄
   end
 end
 
 @adjoint function pairwise(::Euclidean, X::AbstractMatrix, Y::AbstractMatrix; dims=2)
-  D, back = pullback(
-    (X, Y) -> pairwise(SqEuclidean(), X, Y; dims = dims),
-    X,
-    Y,
-  )
+  D, back = pullback((X, Y) -> pairwise(SqEuclidean(), X, Y; dims=dims), X, Y)
   D .= sqrt.(D)
-  return D, Δ -> (nothing, back(Δ ./ (2 .* D))...)
+  function pairwise_pullback(Δ)
+    return (nothing, back(Δ ./ (2 .* (D .+ 1000 * eps(eltype(D)))))...)
```

Comment:
why is there 1000?

Reply:
Seemed to make FiniteDifferences and this implementation line up best, but of course it could be that `FiniteDifferences` is also wrong near the origin...

Reply:
To be blunt, it was hard to get tests to pass otherwise. I'm really just not sure how to get good ground truth values to test against.